packages/server-wallet/src/protocols/ledger-funder.ts

```diff
@@ -0,0 +1,102 @@
+import {BN, checkThat, isSimpleAllocation} from '@statechannels/wallet-core';
+import _ from 'lodash';
+import {Transaction} from 'objection';
+import {Logger} from 'pino';
+
+import {ChainServiceInterface} from '../chain-service';
+import {Channel} from '../models/channel';
+import {LedgerRequest} from '../models/ledger-request';
+import {DBOpenChannelObjective} from '../models/objective';
+import {Store} from '../wallet/store';
+import {WalletResponse} from '../wallet/wallet-response';
+
+export class LedgerFunder {
+  constructor(
+    private store: Store,
+    private chainService: ChainServiceInterface,
+    private logger: Logger,
+    private timingMetrics = false
+  ) {}
+
+  public static create(
+    store: Store,
+    chainService: ChainServiceInterface,
+    logger: Logger,
+    timingMetrics = false
+  ): LedgerFunder {
+    return new LedgerFunder(store, chainService, logger, timingMetrics);
+  }
+
+  /**
+   * Runs the ledger funding protocol
+   *
+   * Returns true if the channel is funded, false otherwise
+   */
+  public async crank(
+    objective: DBOpenChannelObjective,
+    channel: Channel,
+    response: WalletResponse,
+    tx: Transaction
+  ): Promise<boolean> {
+    const ledgerId = objective.data.fundingLedgerChannelId;
+    // sanity check
+    if (!ledgerId) {
+      throw new Error(`LedgerFunder called with objective without a fundingLedgerChannelId`);
+    }
+
+    const ledger = await this.store.getChannel(ledgerId, tx);
+    if (!ledger) return false;
+
+    // are we already funded? if so, return true
+    if (await this.alreadyFunded(channel, ledger, tx)) return true;
+
+    // if we already requested funding return false
+    if (await this.alreadyRequestedFunding(channel.channelId, tx)) return false;
+
+    // otherwise request funding
+    await this.requestLedgerFunding(channel.channelId, ledgerId, tx);
+    return false;
+  }
+
+  private async alreadyFunded(
+    channel: Channel,
+    ledger: Channel,
+    _tx: Transaction
+  ): Promise<boolean> {
+    // for time being ledger must be a direct channel
+    // TODO: in the future check "funding table"
+    if (!ledger.isDirectFunded) return false;
+
+    // ledger is running
+    if (!ledger.isRunning) return false;
+
+    return ledgerFundedThisChannel(ledger, channel);
+  }
+
+  private async requestLedgerFunding(
+    channelId: string,
+    ledgerId: string,
+    tx: Transaction
+  ): Promise<void> {
+    await LedgerRequest.requestLedgerFunding(channelId, ledgerId, tx);
+  }
+
+  private async alreadyRequestedFunding(channelId: string, tx: Transaction): Promise<boolean> {
+    const req = await this.store.getLedgerRequest(channelId, 'fund', tx);
+    return !!req;
+  }
+}
+
+function ledgerFundedThisChannel(ledger: Channel, app: Channel): boolean {
```

Comment:
We could change it to `ledgerFundsApp(...)` or even `doesLedgerFundApp`, if we want to make the query nature clear?

We could also put it on the `Channel`, but I'd probably prefer to see `ledger.funds(app: Channel)`, as the answer depends primarily on data contained within the ledger object. This might not be wise though, given our plans to introduce a funding table. In general, I think it's probably good to make it so that methods on models don't need to reach outside that model.

Reply:
I like `doesLedgerFundApp`.